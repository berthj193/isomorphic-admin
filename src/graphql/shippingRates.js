import gql from 'graphql-tag';

export const getList = gql`query shipping_rates{
  shipping_rates {
    id
    country
    name
    premium
    live
    created_at
  }
}`;

export const getDetails = gql`query shipping_rate(
  $id: ID, $cart_item_count: Int, $country: Country
) {
  shipping_rates(
    id: $id,
    cart_item_count: $cart_item_count,
    country: $country
  ) {
    id
    country
    name
    premium
    live
    created_at
    min_items
    max_items
    shipping_code
    updated_at
    currency
    price_pence
    warehouse {
      id
    }
  }
}`;

export const destroyShippingRate = gql`mutation destroyShippingRate(
  $id: ID!
) {
  destroyShippingRate(
    id: $id
  )
}`;

export const createShippingRate = gql`mutation createShippingRate(
  $warehouse_id: ID!, $country: Country!, $name: String!,
  $live: Boolean, $min_items: Int, $max_items: Int,
  $premium: Boolean, $currency: Currency!, $price_pence: Int!,
  $shipping_code: String!
) {
  createShippingRate(
    warehouse_id: $warehouse_id,
    country: $country,
    live: $live,
    max_items: $max_items,
    min_items: $min_items,
    name: $name,
    premium: $premium,
    currency: $currency,
    price_pence: $price_pence
    shipping_code: $shipping_code
  ) {
    id
  }
}`;

export const updateShippingRate = gql`mutation updateShippingRate(
  $id: ID!, $warehouse_id: ID, $country: Country, $name: String,
  $live: Boolean, $min_items: Int, $max_items: Int,
  $premium: Boolean, $currency: Currency, $price_pence: Int,
  $shipping_code: String
) {
  updateShippingRate(
    id: $id,
    warehouse_id: $warehouse_id,
    country: $country,
    live: $live,
    max_items: $max_items,
    min_items: $min_items,
    name: $name,
    premium: $premium,
    currency: $currency,
    price_pence: $price_pence
    shipping_code: $shipping_code
  ) {
    id
  }
}`;

export const getAutocompleteList = gql`query {
  shipping_rates {
    id
    country
    name
  }
}`;
