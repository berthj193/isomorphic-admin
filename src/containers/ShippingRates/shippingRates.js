import React, { Component } from 'react';
import PropTypes from 'prop-types';
import map from 'lodash/map';
import ItemList from '../../components/list/item-list';
import withStoresManager from '../../components/apolloProviders/withStoresManager';
import Edit from './edit';
import Icon from '../../components/uielements/icon';

import columns from './columns';
import { getList, destroyShippingRate } from '../../graphql/shippingRates';
import DeleteButton from '../../components/uielements/deleteButton';

export class ShippingRates extends Component {
  static propTypes = {
    shipping_rates: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      country: PropTypes.string,
      name: PropTypes.string,
      premium: PropTypes.bool,
      live: PropTypes.bool,
      created_at: PropTypes.string,
    })),
    error: PropTypes.string,
    mutate: PropTypes.func.isRequired,
    refresh: PropTypes.func.isRequired,
  };

  static defaultProps = {
    shipping_rates: [],
  };

  dataSource() {
    return map(this.props.shipping_rates, rate => ({
      key: rate.id,
      id: rate.id,
      country: rate.country,
      name: rate.name,
      premium: <Icon type={rate.premium ? 'check' : 'close'} />,
      live: <Icon type={rate.live ? 'check' : 'close'} />,
      created_at: rate.created_at,
    }));
  }

  deleteShippingRate = id => e => {
    e.preventDefault();
    e.stopPropagation();
    this.props.mutate({
      mutation: destroyShippingRate,
      variables: {
        id,
      },
      messages: ['Legal entity removed successfully'],
    });
  };

  actionColumn = ({ id }) => (
    <ul className="table-action">
      <li>
        <DeleteButton
          tooltipText="Delete shipping rate"
          onConfirm={this.deleteShippingRate(id)}
        />
      </li>
    </ul>
  );

  columns = [
    ...columns,
    {
      title: 'Action',
      dataIndex: '',
      key: 'a',
      render: this.actionColumn,
    },
  ];

  render() {
    return (
      <ItemList
        data={this.dataSource()}
        error={this.props.error}
        title="Shipping rates"
        columns={this.columns}
        detailsComponent={Edit}
        onModalClose={this.onModalClose}
        detailsProps={{ onSubmit: this.handleRefund }}
        buttons={[{
          title: 'New shipping rate',
          link: '/dashboard/shipping_rates/new',
        }]}
        loading={!this.props.shipping_rates && !this.props.error}
      />
    );
  }
}

export default withStoresManager(ShippingRates, { query: getList });
