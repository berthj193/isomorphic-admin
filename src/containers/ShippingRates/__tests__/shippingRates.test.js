import React from 'react';
import { ShippingRates } from '../shippingRates';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import StoreClient from '../../../helpers/client/storeManagerClient';

jest.mock('../../../helpers/client/storeManagerClient');

describe('<ShippingRates />', () => {
  const defaultProps = {
    query: jest.fn(),
    mutate: jest.fn(),
    refresh: jest.fn(),
  };
  const getComponent = (props = {}) => shallow(<ShippingRates {...Object.assign(defaultProps, props)} />);
  const mutate = jest.fn();
  const mockShippingRates = [{
    id: '1',
    country: 'United Kingdom',
    name: 'Shipping rate example',
    premium: true,
    live: false,
    created_at: 'yesterday',
  }];

  beforeEach(() => {
    jest.resetAllMocks();
    StoreClient.query.mockReturnValue(Promise.resolve({ data: { stores: [] } }));
    mutate.mockReturnValue(Promise.resolve({ foo: 'bar' }));
  });

  it('should render properly', () => {
    expect(toJson(getComponent())).toMatchSnapshot();
  });

  it('should map and pass shipping rates to table component', () => {
    const mockOrders = [
      {
        id: '1',
        country: 'United Kingdom',
        name: 'Test',
        premium: true,
        live: true,
        created_at: '2018-05-31 15:43:15 UTC',
      },
    ];
    const component = getComponent();
    component.instance().setState({
      orders: mockOrders,
    });

    expect(toJson(component.update())).toMatchSnapshot();
  });

  it('action column should match snapshot', () => {
    expect(toJson(getComponent().instance().actionColumn(mockShippingRates[0]))).toMatchSnapshot();
  });

  it('should execute mutate function on delete', () => {
    const wrapper = getComponent({ mutate });
    const event = {
      preventDefault: () => null,
      stopPropagation: () => null,
    };
    wrapper.instance().deleteShippingRate(1)(event);
    expect(mutate).toBeCalled();
  });
});
