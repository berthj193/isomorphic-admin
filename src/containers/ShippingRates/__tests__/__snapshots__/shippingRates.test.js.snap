// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShippingRates /> action column should match snapshot 1`] = `undefined`;

exports[`<ShippingRates /> should map and pass shipping rates to table component 1`] = `
<withRouter(ItemList)
  buttons={
    Array [
      Object {
        "link": "/dashboard/shipping_rates/new",
        "title": "New shipping rate",
      },
    ]
  }
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "key": "id",
        "title": "ID",
      },
      Object {
        "dataIndex": "country",
        "key": "country",
        "title": "Country",
      },
      Object {
        "dataIndex": "name",
        "key": "name",
        "title": "Name",
      },
      Object {
        "dataIndex": "premium",
        "key": "premium",
        "title": "Premium",
      },
      Object {
        "dataIndex": "live",
        "key": "live",
        "title": "Live",
      },
      Object {
        "dataIndex": "created_at",
        "key": "created_at",
        "title": "Created At",
      },
      Object {
        "dataIndex": "",
        "key": "a",
        "render": [Function],
        "title": "Action",
      },
    ]
  }
  data={Array []}
  detailsComponent={[Function]}
  detailsProps={
    Object {
      "onSubmit": undefined,
    }
  }
  loading={false}
  title="Shipping rates"
/>
`;

exports[`<ShippingRates /> should render properly 1`] = `
<withRouter(ItemList)
  buttons={
    Array [
      Object {
        "link": "/dashboard/shipping_rates/new",
        "title": "New shipping rate",
      },
    ]
  }
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "key": "id",
        "title": "ID",
      },
      Object {
        "dataIndex": "country",
        "key": "country",
        "title": "Country",
      },
      Object {
        "dataIndex": "name",
        "key": "name",
        "title": "Name",
      },
      Object {
        "dataIndex": "premium",
        "key": "premium",
        "title": "Premium",
      },
      Object {
        "dataIndex": "live",
        "key": "live",
        "title": "Live",
      },
      Object {
        "dataIndex": "created_at",
        "key": "created_at",
        "title": "Created At",
      },
      Object {
        "dataIndex": "",
        "key": "a",
        "render": [Function],
        "title": "Action",
      },
    ]
  }
  data={Array []}
  detailsComponent={[Function]}
  detailsProps={
    Object {
      "onSubmit": undefined,
    }
  }
  loading={false}
  title="Shipping rates"
/>
`;
